device.go

```diff
@@ -92,21 +81,35 @@ func (d *device) InstallUpdate(image io.ReadCloser, size int64) error {
 		return err
 	}
 
-	log.Debugf("Installing update to inactive partition: %s", inactivePartition)
+	b := &BlockDevice{Path: inactivePartition}
 
-	partitionSize, err := d.getPartitionSize(inactivePartition)
-	if err != nil {
+	if bsz, err := b.Size(); err != nil {
+		log.Errorf("failed to read size of block device %s: %v",
+			inactivePartition, err)
 		return err
+	} else if bsz < uint64(size) {
+		log.Errorf("update (%v bytes) is larger than the size of device %s (%v bytes)",
+			size, inactivePartition, bsz)
+		return syscall.ENOSPC
+	}
+
+	w, err := io.Copy(b, image)
+	if err != nil {
+		log.Errorf("failed to write image data to device %v: %v",
+			inactivePartition, err)
 	}
 
-	if size <= partitionSize {
-		if err := writeToPartition(image, size, inactivePartition); err != nil {
-			return err
+	log.Infof("written %v/%v bytes of update to device %v",
+		w, size, inactivePartition)
+
+	if cerr := b.Close(); cerr != nil {
```

Comment:
Should this be a `defer` function? What will happen if there is a `io.Copy()` error?